core/drivers/i2c/atmel_i2c.c

```diff
@@ -0,0 +1,353 @@
+
+	/*
+	 * Since we will configure both CHDIV and CLDIV with the same value
+	 * use 2 * clk
+	 */
+	div = UDIV_ROUND_NEAREST(clk, 2 * I2C_BUS_FREQ) - 3;
+	if (div < 0)
+		div = 0;
+
+	/* CHDIV/CLDIV are on 8 bits, CKDIV on 3 bits */
+	ckdiv = flsi(div >> 8);
+	if (ckdiv > TWIHS_CWGR_CKDIV_MAX) {
+		EMSG("CKDIV value too large");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+	cxdiv = div >> ckdiv;
+
+	if (i2c->sda_hold_time) {
+		/* hold_time = (HOLD + 3) x tperipheral clock */
+		hold = UDIV_ROUND_NEAREST(i2c->sda_hold_time * clk, 1000000000);
+		hold -= 3;
+		if (hold < 0 || hold > TWIHS_CWGR_HOLD_MAX) {
+			EMSG("Incorrect hold value");
+			return TEE_ERROR_BAD_PARAMETERS;
+		}
+
+		cwgr |= hold << TWIHS_CWGR_HOLD_SHIFT;
+	}
+
+	cwgr |= ckdiv << TWIHS_CWGR_CKDIV_SHIFT;
+	/* CHDIV == CLDIV */
+	cwgr |= cxdiv << TWIHS_CWGR_CHDIV_SHIFT;
+	cwgr |= cxdiv;
+	io_write32(i2c->base + TWIHS_CWGR, cwgr);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result atmel_i2c_init_hw(struct atmel_i2c *i2c)
+{
+	/* Unlock TWIHS IP */
+	io_write32(i2c->base + TWIHS_WPMR, TWIHS_WPMR_WPKEY);
+
+	/* Configure master mode */
+	io_write32(i2c->base + TWIHS_CR, TWIHS_CR_SWRST);
+
+	io_write32(i2c->base + TWIHS_CR, TWIHS_CR_SVDIS);
+	io_write32(i2c->base + TWIHS_CR, TWIHS_CR_MSEN);
+
+	return atmel_i2c_init_clk(i2c);
+}
+
+static const struct i2c_ctrl_ops atmel_i2c_ops = {
+	.read = atmel_i2c_read_data,
+	.write = atmel_i2c_write_data,
+	.smbus = atmel_i2c_smbus,
+};
+
+static struct i2c_dev *atmel_i2c_get_dt_i2c(struct dt_driver_phandle_args *a,
+					    void *data, TEE_Result *res)
+{
+	struct i2c_dev *i2c_dev = NULL;
+	struct i2c_ctrl *i2c_ctrl = data;
+
+	i2c_dev = i2c_create_dev(i2c_ctrl, a->fdt, a->phandle_node);
+	if (!i2c_dev) {
+		*res = TEE_ERROR_OUT_OF_MEMORY;
+		return NULL;
+	}
+
+	*res = TEE_SUCCESS;
+	return i2c_dev;
+}
+
+static TEE_Result atmel_i2c_node_probe(const void *fdt, int node,
+				       const void *compat_data __unused)
+{
+	size_t size = 0;
+	const uint32_t *cuint = 0;
+	unsigned int matrix_id = 0;
+	struct i2c_ctrl *i2c_ctrl = NULL;
+	struct atmel_i2c *atmel_i2c = NULL;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	int status = _fdt_get_status(fdt, node);
+
+	if (status != DT_STATUS_OK_SEC)
+		return TEE_SUCCESS;
+
+	atmel_i2c = calloc(1, sizeof(struct atmel_i2c));
+	if (!atmel_i2c)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	i2c_ctrl = &atmel_i2c->i2c_ctrl;
+	i2c_ctrl->ops = &atmel_i2c_ops;
+
+	res = clk_dt_get_by_index(fdt, node, 0, &atmel_i2c->clk);
+	if (res)
+		return res;
+
+	res = matrix_dt_get_id(fdt, node, &matrix_id);
+	if (res)
+		return res;
+
+	if (dt_map_dev(fdt, node, &atmel_i2c->base, &size, DT_MAP_AUTO) < 0)
+		return TEE_ERROR_GENERIC;
+
+	matrix_configure_periph_secure(matrix_id);
+
+	clk_enable(atmel_i2c->clk);
+
+	cuint = fdt_getprop(fdt, node, "i2c-sda-hold-time-ns", NULL);
+	if (cuint)
+		atmel_i2c->sda_hold_time = fdt32_to_cpu(*cuint);
+
+	res = atmel_i2c_init_hw(atmel_i2c);
+	if (res)
+		return res;
```

Comment:
Maybe `clk_disable()` + `free(atmel_i2c)` or error path? But I guess at this point, errors are unrecoverable.

Reply:
I will add an error path !